zaza/openstack/charm_tests/hacluster/tests.py

```diff
@@ -69,3 +74,98 @@ def test_920_put_in_maintenance(self):
 
         self._toggle_maintenance_and_wait('true')
         self._toggle_maintenance_and_wait('false')
+
+
+class HaclusterScalebackTest(HaclusterBaseTest):
+    """hacluster scaleback tests."""
+
+    _PRINCIPLE_APP_NAME = 'keystone'
+    _HACLUSTER_APP_NAME = 'hacluster'
+    _HACLUSTER_CHARM_NAME = 'hacluster'
+
+    def test_930_scaleback(self):
+        """Remove a unit, recalculate quorum and add a new one."""
+        principle_units = zaza.model.get_status().applications[
+            self._PRINCIPLE_APP_NAME]['units']
+        self.assertEqual(len(principle_units), 3)
+        doomed_principle = sorted(principle_units.keys())[0]
+        doomed_unit = juju_utils.get_subordinate_units(
+            [doomed_principle], charm_name=self._HACLUSTER_CHARM_NAME)[0]
+
+        logging.info('Pausing unit {}'.format(doomed_unit))
+        zaza.model.run_action(
+            doomed_unit,
+            'pause',
+            raise_on_failure=True)
+        logging.info('OK')
+
+        logging.info('Removing {}'.format(doomed_principle))
+        zaza.model.destroy_unit(
+            self._PRINCIPLE_APP_NAME,
+            doomed_principle,
+            wait_disappear=True)
+        logging.info('OK')
+
+        expected_states = {
+            self._HACLUSTER_APP_NAME: {
+                "workload-status": "blocked",
+                "workload-status-message":
+                    "Insufficient peer units for ha cluster (require 3)"
+            },
+            self._PRINCIPLE_APP_NAME: {
+                "workload-status": "blocked",
+                "workload-status-message": "Database not initialised",
+            },
+
+            # NOTE(lourot): these applications are present in the zaza bundles
+            # used to test mysql-router, against which we also run this test:
+            'glance': {
+                "workload-status": "waiting",
+                "workload-status-message": "Incomplete relations: identity",
+            },
+            'neutron-api': {
+                "workload-status": "waiting",
+                "workload-status-message": "Incomplete relations: identity",
+            },
+            'nova-cloud-controller': {
+                "workload-status": "waiting",
+                "workload-status-message": "Incomplete relations: identity",
+            },
+        }
+        zaza.model.wait_for_application_states(states=expected_states)
+        zaza.model.block_until_all_units_idle()
+        logging.info('OK')
+
+        logging.info('Adding a hacluster unit')
+        zaza.model.add_unit(self._PRINCIPLE_APP_NAME, wait_appear=True)
+        expected_states = {
+            self._HACLUSTER_APP_NAME: {
+                "workload-status": "active",
+                "workload-status-message": "Unit is ready and clustered"
+            },
+
+            # NOTE(lourot): these applications remain waiting/blocked after
```

Comment:
Same as above, we should handle partial `wait_for_application_states` better if this is necessary.